pkg/tracing/tracing.go

```diff
@@ -0,0 +1,164 @@
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+const (
+	ServiceName = "gateway-api-inference-extension"
+
+	envOTELTracingEnabled      = "OTEL_TRACING_ENABLED"
+	envOTELExporterEndpoint    = "OTEL_EXPORTER_OTLP_ENDPOINT"
+	envOTELServiceName         = "OTEL_SERVICE_NAME"
+	envOTELSamplingRate        = "OTEL_SAMPLING_RATE"
+
+	OperationGatewayRequest       = "gateway.ext_proc.request"
+	OperationRequestOrchestration = "gateway.request_orchestration"
+	OperationScheduling           = "gateway.scheduling"
+
+	AttrGenAIRequestModel = "gen_ai.request.model"
+	AttrOperationOutcome  = "operation.outcome"
+
+	OutcomeSuccess = "success"
+	OutcomeError   = "error"
+)
+
+type Config struct {
+	Enabled          bool
+	ExporterEndpoint string
+	SamplingRate     float64
+	ServiceName      string
+}
+
+func NewConfigFromEnv() *Config {
+	config := &Config{
+		Enabled:          false,
+		ExporterEndpoint: "http://localhost:4317",
+		SamplingRate:     0.1,
+		ServiceName:      ServiceName,
+	}
+
+	if enabled := os.Getenv(envOTELTracingEnabled); enabled != "" {
+		if enabledBool, err := strconv.ParseBool(enabled); err == nil {
+			config.Enabled = enabledBool
+		}
+	}
+
+	if endpoint := os.Getenv(envOTELExporterEndpoint); endpoint != "" {
+		config.ExporterEndpoint = endpoint
+	}
+
+	if serviceName := os.Getenv(envOTELServiceName); serviceName != "" {
+		config.ServiceName = serviceName
+	}
+
+	if samplingRate := os.Getenv(envOTELSamplingRate); samplingRate != "" {
+		if rate, err := strconv.ParseFloat(samplingRate, 64); err == nil {
+			config.SamplingRate = rate
+		}
+	}
+
+	return config
+}
+
+// Initialize sets up OpenTelemetry tracing with the given configuration.
+// It always sets up context propagation, even if tracing is disabled.
+func Initialize(ctx context.Context, config *Config) (func(), error) {
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		propagation.Baggage{},
+	))
+
+	if !config.Enabled {
+		// Return a no-op shutdown function if tracing is disabled
+		return func() {}, nil
+	}
+
+	conn, err := grpc.DialContext(ctx, config.ExporterEndpoint,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithBlock(),
+		grpc.WithTimeout(5*time.Second),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create gRPC connection to collector: %w", err)
+	}
+
+	exporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithGRPCConn(conn))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create OTLP trace exporter: %w", err)
+	}
+
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			semconv.ServiceNameKey.String(config.ServiceName),
+			semconv.ServiceVersionKey.String("1.0.0"),
```

Comment:
Does it make sense to use [debug/buildinfo ](https://pkg.go.dev/debug/buildinfo)here?